dev_guide/cicd.adoc

```diff
@@ -0,0 +1,254 @@
+Coupled with the additional Jenkins build steps that allow verifications of OpenShift `*Services*` and `*ReplicationContorllers*`, the building blocks to validate changes prior to promoting them via tagging take shape.
+
+More on this later when we dive into the more precise examples of the use of these tools.
+
+[[portability]]
+== Portability
+
+Building upon how various OpenShift API Objects and Docker Image count as elements of your application (in addition to the classic binary executables / consumables produced by traditional programming
+language files), once you've decided that a usable version of the application is available for a given stage of you pipeline, the next consideration is how can you import those API Objects and Images into the
+corresponding staging environments of your pipeline.  And in the case of API Objects (much more *mutable* than Images), you quite possibly will need to transform them in some fashion as you propagate them
+through the pipeline to varying environments.
+
+=== Images
+
+First, the simpler of the two (the immutability tenet and the common Docker based interfaces help once again) - Images:
+- multiple `*Projects*` can access the same `*ImageStreams*`within a OpenShift Internal Image Registry
+- multilple `*Clusters*` can access the same external Docker registries
+
+The CLI commands like `oc tag`, `oc import-image` noted in the Versioning section are tools for propagating and tranferring in addition to versioning.
+
+See the specific examples and scenarios below.
+
+=== API Objects
+
+You can get the exact state that instant of an API object, via `oc get`, or obtain versions of the object suitable for recreation (i.e. the *specification* elements noted earlier), with dynamic, environment specific data removed,
+via `oc export`.  However, more likely, the `oc export` command, will often be more appropriate since it can filter out the dynamic, specific state of an API Object, and only retrieve the configuartion and specification aspects.
+The `oc export` command could also prove useful in structuring the API Objects for your application since it can facilitate the creation of re-usable templates.  And if needed, it has flexibility if there allow the retrieveal of
+specific *runtime* elements of an API Object that are reusable across the different runtime environments of your `*CI/CD*` pipeline.
+
+And when you check out said JSON/YAML from the `*SCM*` and run `oc create` (in the case of objects) or `oc process` (in the case of templates) to create in a given environment of you `*CI/CD*` pipeline, there are a
+plethora of commands to edit objects in the API server once they are created there if small transformations are needed when the objects migrate from stage to stage in your `*CI/CD*` pipeline.
```

Comment:
suggest: once they are created. These commands enable small transformations which may be necessary when migrating API objects between stages of your CI/CD pipeline.